fastavro/_writer.pyx

```diff
@@ -449,12 +449,22 @@ cpdef write_union(bytearray fo, datum, schema):
             raise ValueError(msg)
     else:
         pytype = type(datum)
+        best_match_index = -1
+        most_fields = -1
         for index, candidate in enumerate(schema):
             if validate(datum, candidate):
-                break
-        else:
+                if extract_record_type(candidate) == 'record':
+                    fields = len(candidate['fields'])
+                else:
+                    best_match_index = index
+                    break
+                if fields > most_fields:
```

Comment:
Can you move this `if` clause to be after `fields = len(candidate['fields'])` since it only applies for that case?